extensions/chathistory.md

```diff
@@ -0,0 +1,106 @@
+
+When a client with the above mentioned capabilities requests `chathistory` content from the server (using the `CHATHISTORY` command outlined below), the server SHOULD return to the client a single batch containing raw IRC lines starting with and exluduing the `start` parameter and up to and including the message at the `end` parameter. The raw IRC lines SHOULD be formatted and returned to the client as they were originally, with the addition of the above capability tags.
+
+The `server-time` SHOULD be the time at which the message was received from the IRC server and the `batch id` a unique ID for the entire batch. `draft/label` SHOULD be included and MUST be a unique ID used to identify the `chathistory` request and any replies. A `draft/msgid` to identify each individual message MUST be the `draft/msgid` included when each message was first received from the IRC server.
+
+### `CHATHISTORY` Command
+`CHATHISTORY` content can be requested by the client to the server by sending the `CHATHISTORY` command to the server. A `batch` MUST be returned by the server. If no content exists to return, an empty batch SHOULD be returned to avoid the client waiting for a reply. Command support is sent to the client as the RPL_ISUPPORT 005 numeric `:irc.host 005 nick CHATHISTORY=max_messages :are supported by this server`
+
+Both the `limit` and `max_messages` MUST be integers, and `max_messages` MUST be greater than or equal to zero. The client SHOULD NOT request a `limit` with an absolute value greater than the `max_messages` parameter sent in the command. If the `limit` is 0 or no `limit` is provided, the number of messages equal to `max_messages` MUST BE returned.
+
+Both the `start` and `end` parameters support `draft/msgid` and `timestamp`. If the number of lines between the `start` and `end` parameters exceeds the `max_messages`, the server SHOULD return a number of lines equal to the `max_messages` and the appropriate warning as described below. A`limit` or `max_messages` of 0 indicates that no limit or maximum exists.
+
+The `target` parameter specifies a single channel or query from which history SHOULD be retrieved. Wildcards or multiple targets are not supported.
+
+#### Subcommand
+
+The following subcommands are used to describe how the server should return messages relative to the `timestamp(s)` or `draft/msgid(s)` given.
+
+`BEFORE` get up to `limit` number of messages before the given `timestamp` or `draft/msgid`. The `limit` MUST BE positive.
+
+`AFTER` get up to `limit` number of messages after the given `timestamp` or `draft/msgid`. The `limit` MUST BE positive.
+
+`LATEST` get the most recent (up to `limit`) messages that have been sent since the given `timestamp` or `draft/msgid`. The `limit` MUST BE positive.
+
+`AROUND` get a number of messages before and after the `timestamp` or `draft/msgid`. The `limit` MUST BE positive.
```

Comment:
What issue does `AROUND` solve? It's not obvious to me when I would want to use this command.

Reply:
It seems like an odd case where a client requests "around 50" but the server only has 20 above and 30 after.

Reply:
From IRC:

Sunday, February 24th, 2019
13:45:35 <@jwheare> xPaw: AROUND is to enable stuff like search in clients. Search index returns a matching id/timestamp, useful to show context around it.
13:45:47 <@jwheare> Or like a highlights browser
14:11:17 <@xPaw> are you saying it should request AROUND for each search match?
16:25:11 <@jwheare> xPaw: it could do, possibly on demand
16:25:24 <@jwheare> (lazily)
16:26:35 <@xPaw> That would be ridiculously inefficient. Are you talking about server side searching in this case? (Which isn't defined in that spec)
16:28:27 <@jwheare> not necessarily that inefficient. search was just an example. anything that lets you jump back to a point in time in a client. it would be useful to load in context around that point